### PR TITLE
#21, Context for mochaOpts.requires

### PR DESCRIPTION
**Pull request by monolithed**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -45,4 +46,74 @@ Options will be passed to the Mocha instance. See the full list of Mocha options
...
+**wdio.conf.js**
+
+```js
+{
+  suites: {
+    login: ['tests/login/*.js']
+  },
+
+  mochaOpts: {
+    require: './hooks/mocha.js'
```
Comment (christian-bromann) [inline README.md]: This should be an array in case you want to require multiple modules.

File: lib/adapter.js [hunk 2]
```diff
@@ -161,16 +164,17 @@ class MochaAdapter {
         return message
     }
 
-    requireExternalModules (compilers = [], requires = []) {
-        compilers.concat(requires).forEach((mod) => {
-            mod = mod.split(':')
-            mod = mod[mod.length - 1]
+    requireExternalModules (modules, context) {
+        modules.forEach(module => {
```
Comment (christian-bromann) [inline lib/adapter.js]: How should this work? `modules` is an array of arrays

**Comment by christian-bromann** [comment]
@monolithed two comments, also could you please add an integration test that runs a suite with a compiler (or multiple ones) and required modules?

File: README.md (hunk 1 above)
Comment (monolithed) [inline README.md]: Why? `require` and `compilers` are not the same in Mocha

File: lib/adapter.js (hunk 2 above)
Comment (monolithed) [inline lib/adapter.js]:
```js
let {require, requires, compilers = []} = mochaOpts

this.requireExternalModules([require || requires, ...compilers]
```

I believe that `require` is a string value, not array.

Comment (christian-bromann) [inline lib/adapter.js]: since you can set multiple requires in mocha this should at least support arrays

Comment (christian-bromann) [inline lib/adapter.js]: even if require is a string, compilers is definitely an array too

**Comment by monolithed** [comment]
> could you please add an integration test that runs a suite with a compiler (or multiple ones) and required modules?

```diff
+            it('should load multiple modules', () => {
+                adapter.requireExternalModules(['js:module', './lib/myModule'])
+
+                load.calledWith('module')
+                load.calledWith('/mypath/lib/myModule')
+            })
```

Looks like what you want. Or could you provide an example?

**Comment by christian-bromann** [comment]
@monolithed an additional integration test I think is a good idea. See what I mean in https://github.com/webdriverio/wdio-mocha-framework/blob/master/test/tests.spec.js. It runs the whole Mocha adapter with a certain wdio config. That unit test didn't prevent us from shipping a buggy version last time 😉

**Comment by monolithed** [comment]
@christian-bromann 

```
--compilers <ext>:<module>,...       use the given module(s) to compile files
-r, --require <name>                 require the given module
```

I propose to support an identical behavior that my RP does

**Comment by christian-bromann** [comment]
Why? You sometimes need to have multiple compilers or requires. Why should we limit that to 1?

**Comment by christian-bromann** [comment]
See other integrations: https://github.com/sindresorhus/gulp-mocha#require

**Comment by monolithed** [comment]
Because it can be implemented with standard module system )

**Comment by christian-bromann** [comment]
What do you mean?

**Comment by monolithed** [comment]
What I mean:


```js
// config.js
mochaOpts: {
    require: './hooks/mocha.js'
}
```

```js
// ./hooks/mocha.js

require('foo.js')
require('bar.js')
```

**Comment by monolithed** [comment]
And, why not use `compilers`? It does the same you want.

**Comment by christian-bromann** [comment]
I see. So let's take an example, let's say I want to require sinon (https://www.npmjs.com/package/mocha-sinon) and should.js. With your implementation you have to create your own module that requires both. If we support an array format it would be just `require: ['mocha-sinon', 'should.js']`. Since all Mocha adapters also expect an array we should go the same way especially since it isn't a big difference in the implementation.

> And, why not use compilers? It does the same you want.

With your current implementation I also could only apply one single compiler. I haven't seen an example where multiple compilers are required but again we should provide that as it isn't a big difference within the implementation.

**Comment by monolithed** [comment]
> With your current implementation I also could only apply one single compiler. I haven't seen an example where multiple compilers are required but again we should provide that as it isn't a big difference within the implementation.

Hm

```
let {require, requires, compilers = []} = mochaOpts
this.requireExternalModules([require || requires, ...compilers]
```

Is the same:

```
[require || requires].concat(compilers)
```

It works fine )

**Comment by christian-bromann** [comment]
Can you please add that integration test so I have proof of it?

**Comment by monolithed** [comment]
Yep, of course. I need some time )

**Comment by monolithed** [comment]
Can you help me?  I don't understand how to trigger `pre-require` event.

**Comment by christian-bromann** [comment]
@monolithed you don't need to unit test that. More valuable would be an integration test.

**Comment by monolithed** [comment]
@christian-bromann, please see my last commit

**Comment by christian-bromann** [comment]
@monolithed 👍  for the tests but the `options` method is obsolete. Like I said please don't unit test each line of code. A proper integration test is way more useful.

**Comment by monolithed** [comment]
> `options` method is obsolete. 
What do you mean?

> Like I said please don't unit test each line of code. A proper integration test is way more useful.

**Comment by monolithed** [comment]
> Like I said please don't unit test each line of code. A proper integration test is way more useful.

I really don't understand what does mean integration test. Could you please fix one yourself? I believe so we'll save our time :simple_smile:

**Comment by christian-bromann** [comment]
You've rewritten it just for the purpose of making it unit testable which is fine but not necessary here. An integration test would cover those code path and are much more valueable.

> I really don't understand what does mean integration test.

Please see this test: https://github.com/webdriverio/wdio-mocha-framework/blob/master/test/retry.spec.js#L22-L27 . It runs the whole Mocha adapter with a custom wdio.config and a faked WebdriverIO browser global. You need to create a new test spec, a wdio config that has mochaOpts with compiler and require option and a simple mocha test file that runs a simple test like [this](https://github.com/webdriverio/wdio-mocha-framework/blob/master/test/fixtures/sample.spec.js) one. Just make some random compiler and require modules that set some globals and check in these tests if the globals are set.

**Comment by monolithed** [comment]
@christian-bromann,  please see test/options.spec.js, test/fixtures/tests.options.spec.js. it seems working fine ))

**Comment by christian-bromann** [comment]
@monolithed thanks!

Couple of things:
- Mocha remembers which tests get executed so your 2nd test probably doesn't do anything, can you check?
- it seems to be that the require option is still expected to be a string but it should be a string array.

**Comment by monolithed** [comment]
> it seems to be that the require option is still expected to be a string but it should be a string array.

Can you explain why? These options will be identical. it would be misleading for people who read Mocha's documentation

**Comment by christian-bromann** [comment]
I don't see that Mocha has documented in a different way .. it allows you to use multiple require arguments like
```sh
$ mocha --require sinon --require should
```

**Comment by monolithed** [comment]
> Mocha remembers which tests get executed so your 2nd test probably doesn't do anything, can you check?

Thanks, fixed

**Comment by monolithed** [comment]
> I don't see that Mocha has documented in a different way .. it allows you to use multiple require arguments like 
```js
$ mocha --require sinon --require should
```

Ok.

**Comment by monolithed** [comment]
@christian-bromann, done :stuck_out_tongue:

**Comment by monolithed** [comment]
@christian-bromann, Is there anything, that I should do before merge? )

**Comment by christian-bromann** [comment]
Nope, LGTM 👍 
Thanks!

**Comment by christian-bromann** [comment]
published with `v0.4.0`

**Comment by monolithed** [comment]
@christian-bromann, thanks, but I forgot to mention that `requireS`  is no longer available :anguished:  
it can be fixed quickly. 
Or you can make a major release and unpublish the last one

**Comment by monolithed** [comment]
Ah, sorry, there's a failback )

```js
        /**
         * rename requires option to stay backwards compatible
         * ToDo remove with next major release
         */
        if (config.mochaOpts && config.mochaOpts.requires) {
            if (Array.isArray(config.mochaOpts.require)) {
                config.mochaOpts.require.push(config.mochaOpts.requires)
            } else {
                config.mochaOpts.require = config.mochaOpts.requires
            }
        }
```

**Comment by monolithed** [comment]
Thanks!

**Pull request merged**